File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if(head == NULL || head->next == NULL) return NULL;

        int count = 0;
        ListNode* Curr = head;

        while(Curr){
            count++;
            Curr = Curr->next;
        }

        // If removing the first node
        if(count == n) return head->next;

        count = count - n;

        Curr = head;
        ListNode* Prev = NULL;
        ListNode* Next = head->next;

        int i = 0;                       

        while(i < count){                 
            Prev = Curr;
            Curr = Curr->next;
            Next = Curr->next;
            i++;                           
        }

        Prev->next = Next;

        return head;
    }
};

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(head == null || head.next == null) return null;

        ListNode curr = head;
        for(int i = 0; i < n; i++){
            curr = curr.next;
        }
         
        if(curr == null) return head.next;

        ListNode temp = head;
        while(curr.next != null){
            curr = curr.next;
            temp = temp.next;
        }

        temp.next = temp.next.next;
        return head;
    }
}